src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -43,10 +45,10 @@ public CSharpConvertLinqQueryToForEachProvider()
         /// <summary>
         /// Finds a node for the span and checks that it is either a QueryExpressionSyntax or a QueryExpressionSyntax argument within ArgumentSyntax.
         /// </summary>
-        protected override QueryExpressionSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+        protected override async Task<QueryExpressionSyntax> FindNodeToRefactor(Document document, TextSpan selection, CancellationToken cancellationToken)
         {
-            var node = root.FindNode(span);
-            return node as QueryExpressionSyntax ?? (node is ArgumentSyntax argument ? argument.Expression as QueryExpressionSyntax : default);
+            var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>() as CSharpRefactoringHelpersService;
+            return await refactoringHelperService.TryGetSelectedNodeAsync<QueryExpressionSyntax>(document, selection, refactoringHelperService.ExtractNodeFromDeclarationAndAssignment<QueryExpressionSyntax>, cancellationToken).ConfigureAwait(false);
```

Comment:
Should `ExtractNodeFromDeclarationAndAssignment` be defined in the service itself (probably an abstract class implementing the interface)? This way, all the TryGetSelectedNodeAsync helpers can default to attempting `ExtractNodeFromDeclarationAndAssignment` on the code path where it is unable to find any node from selection. This way each refactoring does not need to explicitly request such special extraction logic and all refactorings are consistent as well.

Reply:
I moved it to the service (and created both C# and VB implementation) but am not sure about having it as default for following reasons: 
1) Assuming we'll find more extraction Helpers the question will be which one to use as default.
2) The extraction is done always even when the original node would match. That's why it returns the given node for cases nothing was extracted because the only the result is tested. Might change this, would add one check to each return-place but nbd.

All in all, I'd leave it as it (with default being Identity extractor) and only after we recognize this one / some other one is used almost exclusively set it as default & make appropriate changes.

Reply:
Ok, convinced below.